### PR TITLE
Explanation/development process

### LINKED ISSUE #113: Write explanation "Ubuntu development process"

### PR DESCRIPTION
**Pull request by dviererbe**
> [!IMPORTANT]
> This PR is targeted against #53 for better readability of the diff. **Set the base to the `2.0-preview` branch before merging!**

This PR adds the initial content for the explanation article of the Ubuntu development process.

*(This PR is based on #54, fixed the spelling mistake in the branch name and rebased onto `explanation/merges-and-syncs`)*

### CODE REVIEW

File: docs/explanation/development-process.rst
Comment (dviererbe) [inline docs/explanation/development-process.rst]: > It's not quite clear what the definition of a freeze is here. I think we could be a bit more explicit about what we mean by "freeze", to avoid using the term without making it really clear what the term means.

_Originally posted by @s-makin in https://github.com/canonical/ubuntu-packaging-guide/pull/54#discussion_r1624269513_

File: docs/explanation/development-process.rst
Comment (dviererbe) [inline docs/explanation/development-process.rst]: @s-makin what do you think about:

```suggestion
During development, caution is increasingly exercised in making changes
to Ubuntu to ensure a stable state is reached in time for the final release
date. Modifications to the Ubuntu package archive get incrementaly restricted,
effectively freezing the state of the Ubuntu package archive. The
deadlines when these restrictions get enabled are called "freezes". During
freezes, exceptions must be requested to approve changes. See 
:doc:`how to request a freeze exception </how-to/request-freeze-exception>`.
The typical order and length of the various freezes can be seen on the
current Release Schedule, which is usually posted as a Discourse article under
the `"Release" topic <https://discourse.ubuntu.com/c/release/>`_. 

.. note::
    In the past, the Release Schedule was published in the Ubuntu Wiki.
    See for example the `release schedule of Ubuntu 20.04 LTS (Focal Fossa) <https://wiki.ubuntu.com/FocalFossa/ReleaseSchedule>`_.
```

**Edit:** just noticed that there is a space missing in "FocalFossa"

Comment (s-makin) [reply docs/explanation/development-process.rst]: I think this is great, much clearer!

Comment (dviererbe) [reply docs/explanation/development-process.rst]: Thanks, committed with https://github.com/canonical/ubuntu-packaging-guide/pull/62/commits/6425645543a860e8f087b99eeae60b1370c11540

**Comment by dviererbe** [comment]
The failed CI checks are related to https://discourse.ubuntu.com/ and https://discourse.canonical.com/ being down. Nothing for me to action on, but the RTD build fail, which is weird.

EDIT: Failure is caused by:
`/home/docs/checkouts/readthedocs.org/user_builds/canonical-ubuntu-packaging-guide/checkouts/62/docs/explanation/development-process.rst:16: WARNING: term not in glossary: 'seeds'`

**Comment by dviererbe** [comment]
Nice, all checks have passed here too now! 🎉

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]:
```suggestion
See also the article :doc:`/explanation/releases` for more details about the
```

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]:
```suggestion
experiences and ideas and to inspire future projects covering
```

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]: This is internal process info, and as such I think it shouldn't be included. While not a secret, it suggests there are parts of the devel process that community is excluded from (nothing wrong about the fact itself -- we just don't need to advertise it).
```suggestion
```

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]:
```suggestion
Merging with upstream and feature development
```

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]:
```suggestion
:doc:`Merges and syncs from Debian </explanation/debian-merges-and-syncs>`. 
```

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]:
```suggestion
The development of planned projects for the release often begins while merging is
```

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]:
```suggestion
still underway, and the development accelerates once the package archive
```

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]: Unless it's a required term, I suggest using 'milestones' instead of the slightly morbid deadlines.
```suggestion
Stabilisation and milestones (freezes)
--------------------------------------
```

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]: Can we specify _who_ does the restricting here? It would both make the information more complete and also allow us to avoid the use of the passive voice.

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]: Same ^

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]:
```suggestion
freezes, developers must request exceptions to approve changes. See 
```

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]: Just trying to reshuffle it to eliminate passive.
```suggestion
The release team usually posts the current Release Schedule as a Discourse article
under the `"Release" topic <https://discourse.ubuntu.com/c/release/>`_. It shows the typical order and length of the various freezes.
```

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]:
```suggestion
    See, for example, the `release schedule of Ubuntu 20.04 LTS (Focal Fossa) <https://wiki.ubuntu.com/FocalFossa/ReleaseSchedule>`_.
```

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]:
```suggestion
During the release development phase, developers organise testing weeks to focus the
```

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]: Again, who does the disabling, please?

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst:115]: Just curious: prior to this, there are no restrictions on ABI changes?

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]:
```suggestion
* User interface of individual applications that are installed by default
* Appearance of the desktop
* Distribution-specific artwork
* All user-visible strings in the desktop and applications that are installed 
by default
```

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]:
```suggestion
Exceptions to this rule may be considered before the release for significant and
```

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]:
```suggestion
glaring errors or exceptional circumstances.
```

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]:
```suggestion
package archive. The release team no longer accepts changes in the Ubuntu
```

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]:
```suggestion
For the beta release preparation, all uploads are queued and subject to
```

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]:
```suggestion
uploading. Uploads for packages that do not affect images are generally
```

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]:
```suggestion
The Kernel Freeze is the final date for kernel updates because they require several
```

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst:216]:
```suggestion
Non-language-pack translation final date
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]:
```suggestion
This marks the date after which translations for such packages are not
```

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]:
```suggestion
Ubuntu flavours remain in :ref:`FeatureFreeze` because they do not affect
```

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]:
```suggestion
the release team cherry-picks into the
```

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]:
```suggestion
release pocket until the Final Release become candidates for
```

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]:
```suggestion
must reference at least one bug, which is used to track the stable update. 
```

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]:
```suggestion
.. note::
    If you are sure that your upload will be accepted during Final Freeze, you can
    upload directly to the release pocket, but be aware that you have to re-upload
    after Final Release if the upload gets rejected.
```

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst:280]:
```suggestion
Language pack translation last day
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]:
```suggestion
Once the release team declares the :ref:`ReleaseCandidate` ISO stable and names it
the "Final Release", a representative of the team announces it on the 
`ubuntu-announce mailing list <https://lists.ubuntu.com/archives/ubuntu-announce/>`_.
```

File: docs/explanation/development-process.rst
Comment (rkratky) [inline docs/explanation/development-process.rst]:
```suggestion
    See, for example, the `Ubuntu 24.04 LTS (Noble Numbat) release announcement <https://lists.ubuntu.com/archives/ubuntu-announce/2024-April/000301.html>`_.
```

**Review by rkratky — CHANGES_REQUESTED** [review]
Thank you, @dviererbe, great stuff! I left a bunch of comments and suggestions.

About the passive voice: I realize it's not always possible or suitable, but in places where its use can be avoided, it's worth trying. The text is then easier to read, and sometimes it also forces us to provide more specific info instead of "hiding" behind the anonymity of the passive :)

File: docs/explanation/development-process.rst
Comment (dviererbe) [reply docs/explanation/development-process.rst:280]: note: I agree that the word `deadline` should be avoided for obvious reasons, but it is the term that was and is used in Ubuntu release schedules (see for example the [Oracular Oriole Release Schedule](https://discourse.ubuntu.com/t/oracular-oriole-release-schedule/36460) or [Language Pack Translation Deadline (Ubuntu Wiki)](https://wiki.ubuntu.com/LanguagePackTranslationDeadline)).

thought: Maybe it is worth having a discussion with the release team to change the used terminology.

File: docs/explanation/development-process.rst
Comment (dviererbe) [reply docs/explanation/development-process.rst:115]: No (based on the documentation I have read and the changes I have experienced).

Prior FF, it is effectively the wild west in the archive and also the reason why major disruptive archive event are announced in the release schedule (see e.g. https://discourse.ubuntu.com/t/oracular-oriole-release-schedule/36460)

If a change breaks reverse dependencies, this has to be addressed before the final release by patching the issue, rebuilding with the new dependencies or removing the reverse dependency (just a component or an entire package); also known as "proposed-migration" and "+1 Maintenance".

File: docs/explanation/development-process.rst
Comment (dviererbe) [reply docs/explanation/development-process.rst]: Archive admins disable the automatic import of changes from Debian

File: docs/explanation/development-process.rst
Comment (rkratky) [reply docs/explanation/development-process.rst:280]: Thanks for the context. Please, consider the deadline-related suggestion non-blocking.

File: docs/explanation/development-process.rst
Comment (dviererbe) [reply docs/explanation/development-process.rst:216]: issue: @rkratky this is the wording used in the release schedule (e.g. https://discourse.ubuntu.com/t/noble-numbat-release-schedule/35649), I think that it is confusing if we use a different name

Comment (dviererbe) [reply docs/explanation/development-process.rst:216]: I will set this thread for now as resolved to unblock this PR from merging. In exchange I  opened #73.

**Comment by dviererbe** [comment]
> About the passive voice: I realize it's not always possible or suitable, but in places where its use can be avoided, it's worth trying. The text is then easier to read, and sometimes it also forces us to provide more specific info instead of "hiding" behind the anonymity of the passive :)

I will keep this in mind for writing future texts :)

**Pull request merged**
